Firestore/core/src/firebase/firestore/model/mutation.h

```diff
@@ -0,0 +1,148 @@
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_MUTATION_H_
+
+#include <memory>
+
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+#include "Firestore/core/src/firebase/firestore/model/maybe_document.h"
+#include "Firestore/core/src/firebase/firestore/model/precondition.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+/**
+ * Represents a Mutation of a document. Different subclasses of Mutation will
+ * perform different kinds of changes to a base document. For example, a
+ * SetMutation replaces the value of a document and a DeleteMutation deletes a
+ * document.
+ *
+ * In addition to the value of the document mutations also operate on the
+ * version. For local mutations (mutations that haven't been committed yet), we
+ * preserve the existing version for Set, Patch, and Transform mutations. For
+ * local deletes, we reset the version to 0.
+ *
+ * Here's the expected transition table.
+ *
+ * MUTATION           APPLIED TO      RESULTS IN
+ * SetMutation        Document(v3)    Document(v3)
+ * SetMutation        NoDocument(v3)  Document(v0)
+ * SetMutation        null            Document(v0)
+ * PatchMutation      Document(v3)    Document(v3)
+ * PatchMutation      NoDocument(v3)  NoDocument(v3)
+ * PatchMutation      null            null
+ * TransformMutation  Document(v3)    Document(v3)
+ * TransformMutation  NoDocument(v3)  NoDocument(v3)
+ * TransformMutation  null            null
+ * DeleteMutation     Document(v3)    NoDocument(v0)
+ * DeleteMutation     NoDocument(v3)  NoDocument(v0)
+ * DeleteMutation     null            NoDocument(v0)
+ *
+ * For acknowledged mutations, we use the updateTime of the WriteResponse as the
+ * resulting version for Set, Patch, and Transform mutations. As deletes have no
+ * explicit update time, we use the commitTime of the WriteResponse for
+ * acknowledged deletes.
+ *
+ * If a mutation is acknowledged by the backend but fails the precondition check
+ * locally, we return an `UnknownDocument` and rely on Watch to send us the
+ * updated version.
+ *
+ * Note that TransformMutations don't create Documents (in the case of being
+ * applied to a NoDocument), even though they would on the backend. This is
+ * because the client always combines the TransformMutation with a SetMutation
+ * or PatchMutation and we only want to apply the transform if the prior
+ * mutation resulted in a Document (always true for a SetMutation, but not
+ * necessarily for an PatchMutation).
+ */
+class Mutation {
+ public:
+  virtual ~Mutation() {
+  }
+
+  const DocumentKey& key() const {
+    return key_;
+  }
+  const Precondition& precondition() const {
+    return precondition_;
+  }
+
+  // TODO(rsgowman): ApplyToRemoteDocument()
+
+  /**
+   * Applies this mutation to the given MaybeDocument for the purposes of
```

Comment:
Very optional: consider surrounding class names with backticks (`MaybeDocument`) in comments. OTOH, an argument could be made that it makes copy-pasting code between platforms unnecessarily harder.